### ISSUE TITLE
[Bug]: Error detected In C file

### ISSUE THREAD

**Issue by Six6stRINgs** [HiPhish/rainbow-delimiters.nvim#59]
### Neovim version

0.9.4

### Language affected

C

### Description

This problem occurs when I open a C file by [nvimdots](https://github.com/ayamir/nvimdots).
And errors consistently occur as I move the cursor around.
Sometimes the cursor even move itself in a strange way.
Besides, this problem may have associated with the codes.
If I create a new C file and just do "Hello world" print, no error will be reported.

Error Info:
```
Error detected while processing CursorMoved Autocommands for "<buffer=1>":
Error executing lua callback: ...elimiters.nvim/lua/rainbow-delimiters/strategy/local.lua:198: attempt to call method 'range' (a nil value)
stack traceback:
        ...elimiters.nvim/lua/rainbow-delimiters/strategy/local.lua:198: in function 'update_local'
        ...elimiters.nvim/lua/rainbow-delimiters/strategy/local.lua:223: in function 'fn'
        ...m/share/nvim/runtime/lua/vim/treesitter/languagetree.lua:345: in function 'for_each_tree'
        ...elimiters.nvim/lua/rainbow-delimiters/strategy/local.lua:222: in function 'local_rainbow'
        ...elimiters.nvim/lua/rainbow-delimiters/strategy/local.lua:277: in function <...elimiters.nvim/lua/rainbow-delimiters/strategy/local.lua:273>
```

**Comment by HiPhish** [comment]
Does it happen in any C file? Does it happen with the included test file (`test/highlight/regular.c`). If you are using a premade configuration please try without; these configurations add all sorts of stuff and there is no way I can troubleshoot that.

The error message says that the method `range` does not exist on the `TSTree` object. So chances are that this is some sort of type error.

**Comment by Danielkonge** [comment]
I have fixed this error in #54 already (discovered it when adding types since I got a type warning). In the local strategy there is a line:
```lua 
log.debug("Did not build any matches Stack for tree '%s'", vim.inspect(tree:range()))
```
that should have been 
```lua
log.debug("Did not build any matches Stack for tree '%s'", vim.inspect(tree:root():range()))
```

If we want to quickly update just these small problems, I can make a smaller cleanup pull request with small fixes first?

**Comment by Six6stRINgs** [comment]
> I have fixed this error in #54 already (discovered it when adding types since I got a type warning). In the local strategy there is a line:
> 
> ```lua
> log.debug("Did not build any matches Stack for tree '%s'", vim.inspect(tree:range()))
> ```
> 
> that should have been
> 
> ```lua
> log.debug("Did not build any matches Stack for tree '%s'", vim.inspect(tree:root():range()))
> ```
> 
> If we want to quickly update just these small problems, I can make a smaller cleanup pull request with small fixes first?
> 
> Update: Since it wasn't much work, I made a smaller pull request including this fix now.


I do change this line in the local.lua, but error still detected with different info
```
E5108: Error executing lua: ...y/nvim-tree.lua/lua/nvim-tree/actions/node/open-file.lua:220: CursorMoved Autocommands for "<buffer=6>": Vim(append):Error executing 
lua callback: vim/inspect.lua:0: attempt to index a number value
stack traceback:
        vim/inspect.lua: in function 'inspect'
        ...elimiters.nvim/lua/rainbow-delimiters/strategy/local.lua:217: in function 'update_local'
        ...elimiters.nvim/lua/rainbow-delimiters/strategy/local.lua:244: in function 'fn'
        ...m/share/nvim/runtime/lua/vim/treesitter/languagetree.lua:345: in function 'for_each_tree'
        ...elimiters.nvim/lua/rainbow-delimiters/strategy/local.lua:243: in function 'local_rainbow'
        ...elimiters.nvim/lua/rainbow-delimiters/strategy/local.lua:304: in function <...elimiters.nvim/lua/rainbow-delimiters/strategy/local.lua:300>
        [C]: in function 'nvim_set_current_win'
        ...y/nvim-tree.lua/lua/nvim-tree/actions/node/open-file.lua:220: in function 'set_current_win_no_autocmd'
        ...y/nvim-tree.lua/lua/nvim-tree/actions/node/open-file.lua:284: in function 'open_in_new_window'
        ...y/nvim-tree.lua/lua/nvim-tree/actions/node/open-file.lua:334: in function 'fn'
        ...share/nvim/site/lazy/nvim-tree.lua/lua/nvim-tree/api.lua:170: in function 'edit'
        ...share/nvim/site/lazy/nvim-tree.lua/lua/nvim-tree/api.lua:180: in function 'fn'
        ...share/nvim/site/lazy/nvim-tree.lua/lua/nvim-tree/api.lua:49: in function <...share/nvim/site/lazy/nvim-tree.lua/lua/nvim-tree/api.lua:46>
stack traceback:
        [C]: in function 'nvim_set_current_win'
        ...y/nvim-tree.lua/lua/nvim-tree/actions/node/open-file.lua:220: in function 'set_current_win_no_autocmd'
        ...y/nvim-tree.lua/lua/nvim-tree/actions/node/open-file.lua:284: in function 'open_in_new_window'
        ...y/nvim-tree.lua/lua/nvim-tree/actions/node/open-file.lua:334: in function 'fn'
        ...share/nvim/site/lazy/nvim-tree.lua/lua/nvim-tree/api.lua:170: in function 'edit'
        ...share/nvim/site/lazy/nvim-tree.lua/lua/nvim-tree/api.lua:180: in function 'fn'
        ...share/nvim/site/lazy/nvim-tree.lua/lua/nvim-tree/api.lua:49: in function <...share/nvim/site/lazy/nvim-tree.lua/lua/nvim-tree/api.lua:46>
```


> Does it happen in any C file? Does it happen with the included test file (`test/highlight/regular.c`). If you are using a premade configuration please try without; these configurations add all sorts of stuff and there is no way I can troubleshoot that.
> 
> The error message says that the method `range` does not exist on the `TSTree` object. So chances are that this is some sort of type error.

I not sure whether I use any premade configuration, how to check it?
By the way, I open the `test/highlight` folder and  run the C/C++ files and some other files resulting no error.
But just now I find out how to trigger this bug -- add #define in the C/C++ file. I also add #define in the `regular.c`, and error occurs as I move the cursor to the bracket.

**Comment by Danielkonge** [comment]
> I do change this line in the local.lua, but error still detected with different info
> 
> ```
> E5108: Error executing lua: ...y/nvim-tree.lua/lua/nvim-tree/actions/node/open-file.lua:220: CursorMoved Autocommands for "<buffer=6>": Vim(append):Error executing 
> lua callback: vim/inspect.lua:0: attempt to index a number value
> stack traceback:
>         vim/inspect.lua: in function 'inspect'
>         ...elimiters.nvim/lua/rainbow-delimiters/strategy/local.lua:217: in function 'update_local'
>         ...elimiters.nvim/lua/rainbow-delimiters/strategy/local.lua:244: in function 'fn'
>         ...m/share/nvim/runtime/lua/vim/treesitter/languagetree.lua:345: in function 'for_each_tree'
>         ...elimiters.nvim/lua/rainbow-delimiters/strategy/local.lua:243: in function 'local_rainbow'
>         ...elimiters.nvim/lua/rainbow-delimiters/strategy/local.lua:304: in function <...elimiters.nvim/lua/rainbow-delimiters/strategy/local.lua:300>
>         [C]: in function 'nvim_set_current_win'
>         ...y/nvim-tree.lua/lua/nvim-tree/actions/node/open-file.lua:220: in function 'set_current_win_no_autocmd'
>         ...y/nvim-tree.lua/lua/nvim-tree/actions/node/open-file.lua:284: in function 'open_in_new_window'
>         ...y/nvim-tree.lua/lua/nvim-tree/actions/node/open-file.lua:334: in function 'fn'
>         ...share/nvim/site/lazy/nvim-tree.lua/lua/nvim-tree/api.lua:170: in function 'edit'
>         ...share/nvim/site/lazy/nvim-tree.lua/lua/nvim-tree/api.lua:180: in function 'fn'
>         ...share/nvim/site/lazy/nvim-tree.lua/lua/nvim-tree/api.lua:49: in function <...share/nvim/site/lazy/nvim-tree.lua/lua/nvim-tree/api.lua:46>
> stack traceback:
>         [C]: in function 'nvim_set_current_win'
>         ...y/nvim-tree.lua/lua/nvim-tree/actions/node/open-file.lua:220: in function 'set_current_win_no_autocmd'
>         ...y/nvim-tree.lua/lua/nvim-tree/actions/node/open-file.lua:284: in function 'open_in_new_window'
>         ...y/nvim-tree.lua/lua/nvim-tree/actions/node/open-file.lua:334: in function 'fn'
>         ...share/nvim/site/lazy/nvim-tree.lua/lua/nvim-tree/api.lua:170: in function 'edit'
>         ...share/nvim/site/lazy/nvim-tree.lua/lua/nvim-tree/api.lua:180: in function 'fn'
>         ...share/nvim/site/lazy/nvim-tree.lua/lua/nvim-tree/api.lua:49: in function <...share/nvim/site/lazy/nvim-tree.lua/lua/nvim-tree/api.lua:46>
> ```

This error seems to be because you are somehow calling an autocommand in the nvim-tree buffer, [which shouldn't happen](https://github.com/nvim-tree/nvim-tree.lua/blob/80cfeadf179d5cba76f0f502c71dbcff1b515cd8/lua/nvim-tree/actions/node/open-file.lua#L217C1-L217C1). Thus `vim.inspect` doesn't get the type of input it expects here. This seems more like an error in nvim-tree, but I am not sure exactly how to best solve it. 

> I'm not sure whether I use any premade configuration, how to check it? 

If you use nvimdots, then you are using a premade configuration. 

> By the way, I open the `test/highlight` folder and run the C/C++ files and some other files resulting no error. But just now I find out how to trigger this bug -- add #define in the C/C++ file. I also add #define in the `regular.c`, and error occurs as I move the cursor to the bracket.

What do you mean by "error occurs as I move the cursor to the bracket"? What bracket? With the given information about adding a `#define` I wasn't able to reproduce this.

**Comment by Six6stRINgs** [comment]
> What do you mean by "error occurs as I move the cursor to the bracket"? What bracket? With the given information about adding a `#define` I wasn't able to reproduce this.

If I move to the line that has `{}` , the error will be reported.
As you could't reproduce the problem, I assume that maybe it's caused by a conflict with other plug or  some premade configuration.

**Comment by Danielkonge** [comment]
> If I move to the line that has `{}` , the error will be reported.

I could still not reproduce it.

> As you could't reproduce the problem, I assume that maybe it's caused by a conflict with other plug or some premade configuration.

Yes, I think it is something specific to the nvimdots most likely (if it was nvim-tree I think other people would have experienced this before). If I had to guess, it has something to do with the events they use in the setup, but I am not sure about this.

**Comment by ayamir** [comment]
Hello, I find this bug can be reproduced stably by following steps: https://github.com/ayamir/nvimdots/issues/1065#issuecomment-1816555151

**Comment by Danielkonge** [comment]
> Hello, I find this bug can be reproduced stably by following steps: [ayamir/nvimdots#1065 (comment)](https://github.com/ayamir/nvimdots/issues/1065#issuecomment-1816555151)

I was able to reproduce this with the given steps. Note: This only errors in Neovim 0.9.4, not in the latest version of 0.10.

I will try to see if I can figure out a solution.

**Comment by Danielkonge** [comment]
I have found what causes the error, but now I have noticed another problem.

The error seems to be caused by some problem with `vim.inspect`, where it can't handle the output of `tree:root():range()`. This is easy enough to fix, by changing 
```lua
vim.inspect(tree:root():range())
```
to
```lua
vim.inspect( {tree:root():range()} )
```

Now we don't get any error anymore with this fix, but rainbow-delimiters still doesn't work as expected in a c file with `#define MACRO 0` included, because this seems to mess with the treesitter parser for c. Removing `#define MACRO 0` and gets rainbow-delimiters working, and when it works, putting the macro back in the code doesn't break it. I will try to see if I can get it to work from the start without this annoying workaround.
